# Part I: Theory Questions

## MLE 
### 2. Fill the blanks with T (True) or F (False) for the statements given above:


• Maximum likelihood estimation provides not only point estimation, but a
distribution information of the parameters estimated. (T)

• Maximum likelihood and Bayesian approachs for parameter estimation perform well with low-dimensional dataset with many training examples while
their performance is bad on high-dimensional dataset with few training examples. (T)


## Naive Bayes
### 1. Consider that you are given the dataset in the table above consisting of boolean variables x, y and z and a single boolean output variable C. Suppose that the Naive Bayes classifier is going to be used.


• Specify the value of P(C = 1|x = 1, y = 1, z = 0). Show your solution step
by step.


P(C=1|x=1,y=1,z=0) = P(x=1|C=1)*P(y=1|C=1)*P(z=0|C=1)*P(C=1) = 1/2 * 1/4 * 1/2 * 1/2 = 1/16

• Specify the value of P(C = 0|x = 1, y = 1). Show your solution step by step.


P(C=0|x=1,y=1) = P(x=1|C=0)*P(y=1|C=0)*P(C=0) = 1/2 * 1/2 * 1/2 = 1/8

Now suppose that the Joint Naive Bayes classifier is used for the options below:

• Specify the value of P(C = 1|x = 1, y = 1, z = 0). Show your solution step
by step

x=1, y=1, z=0  C=0  P(C=1|x=1,y=1,z=0) = 0

• Specify the value of P(C = 0|x = 1, y = 1). Show your solution step by step.

x=1,y=1,C=0   x=1,y=1,C=1  P(C=0|x=1,y=1)= 1/2

# PART II: Detection of Fake News

##   INTRODUCTIONS


What is a classifier?

A classifier is a machine learning model that is used to discriminate different objects based on certain features.

Principle of Naive Bayes Classifier:

A Naive Bayes classifier is a probabilistic machine learning model that’s used for classification task. The crux of the classifier is based on the Bayes theorem. Naive Bayes algorithms are mostly used in sentiment analysis, spam filtering, recommendation systems etc. They are fast and easy to implement but their biggest disadvantage is that the requirement of predictors to be independent. In most of the real life cases, the predictors are dependent, this hinders the performance of the classifier.

## Naïve Bayes



With a multinomial event model, samples (feature vectors) represent the frequencies with which certain events have been generated by a multinomial ${\displaystyle (p_{1},\dots ,p_{n})}$ where ${\displaystyle p_{i}}$ is the probability that event i occurs (or $K$ such multinomials in the multiclass case). A feature vector ${\displaystyle \mathbf {x} =(x_{1},\dots ,x_{n})}$ is then a histogram, with ${\displaystyle x_{i}}$ counting the number of times event $i$ was observed in a particular instance. This is the event model typically used for document classification, with events representing the occurrence of a word in a single document (e.g Bag Of Words assumption). The likelihood of observing a histogram $x$ is given by

${\displaystyle p(\mathbf {x} \mid C_{k})={\frac {(\sum _{i}x_{i})!}{\prod _{i}x_{i}!}}\prod _{i}{p_{ki}}^{x_{i}}}$

The multinomial naive Bayes classifier becomes a linear classifier when expressed in log-space:[2]

${\displaystyle {\begin{aligned}\log p(C_{k}\mid \mathbf {x} )&\varpropto \log \left(p(C_{k})\prod _{i=1}^{n}{p_{ki}}^{x_{i}}\right)\\&=\log p(C_{k})+\sum _{i=1}^{n}x_{i}\cdot \log p_{ki}\\&=b+\mathbf {w} _{k}^{\top }\mathbf {x} \end{aligned}}}$

where ${\displaystyle b=\log p(C_{k})}$ and ${\displaystyle w_{ki}=\log p_{ki}}$





### Laplace Smoothing in Naïve Bayes

Let $c$ refer to a class (such as Positive or Negative), and let $w$ refer to a token or word.

The maximum likelihood estimator for $P(w|c)$ is

$$   \frac{count(w,c)}{count(c)}= \frac{counts\,w\,in\,class\,c}{counts\,of\,words\,in\,class\,c}$$


This estimation of $P(w|c)$ could be problematic since it would give us probability $0$ for documents with unknown words. A common way of solving this problem is to use Laplace smoothing.

Let $V$ be the set of words in the training set, add a new element $UNK$ (for unknown) to the set of words.

Define

$$   P(w|c) = \frac{count(w,c)+1}{count(c)+|V|+1}$$

where $V$ refers to the vocabulary (the words in the training set).

In particular, any unknown word will have probability

$$   P(w|c) = \frac{1}{count(c)+|V|+1}$$


## Approach

### Training the Naive Bayes Model

In [1]:
import numpy as np 
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
from NaiveBayes import naive_bayes
from test import run_test, All_Results
from utility import understanding_data

 First of all we have to call desired libraries.
 
 Numpy is a python library which provides matrix operations.
 
 Pandas stands for “Python Data Analysis Library” a must library for data analysis.
 
 The CountVectorizer function is a real game changer when it comes to implementing the naive bayes algorithm. Here is more information about CountVectorization https://towardsdatascience.com/natural-language-processing-count-vectorization-with-scikit-learn-e7804269bb5e

And then we got our stop words and my own libraries
 

 
 [1_YEJf9BQQh0ma1ECs6x_7yQ.png](attachment:1_YEJf9BQQh0ma1ECs6x_7yQ.png)

In [2]:
df = pd.read_csv('fake_news_train.csv',sep=",")

df = df.drop(columns = ['id','author','text'])
df = df[pd.notnull(df['title'])]
#test_size ve random state degistirilebilir

Then we read our fake_news_train.csv file and drop 'id','author' and 'title' because we will not need them in our current experiment. We drop the titles which are null

In [3]:
X_train = df['title']
Y_train = df['label']
dumbx, X_test, dumby, Y_test = (
    train_test_split(df['title'],df['label'],test_size=0.05,random_state=1))

We do not have to split our data to train and test. Therefore I get all data to my X_train and Y_train.
But unfortunately by getting some errors I decided to use train_test_split() function to get my X_test and Y_test datas.

In [4]:
print(X_train)
print(Y_train)

0        FLYNN: Hillary Clinton, Big Woman on Campus - ...
1        15 Civilians Killed In Single US Airstrike Hav...
2        Jackie Mason: Hollywood Would Love Trump if He...
3        Life: Life Of Luxury: Elton John’s 6 Favorite ...
4        Benoît Hamon Wins French Socialist Party’s Pre...
                               ...                        
16635    Rapper T.I.: Trump a ’Poster Child For White S...
16636    N.F.L. Playoffs: Schedule, Matchups and Odds -...
16637    Macy’s Is Said to Receive Takeover Approach by...
16638    NATO, Russia To Hold Parallel Exercises In Bal...
16639                            What Keeps the F-35 Alive
Name: title, Length: 16194, dtype: object
0        0
1        1
2        0
3        1
4        0
        ..
16635    0
16636    0
16637    0
16638    1
16639    1
Name: label, Length: 16194, dtype: int64


In [5]:
print(X_test)
print(Y_test)

15475    In the Shadow of a Fairy Tale - The New York T...
15645    Get Ready For Civil Unrest: Survey Finds That ...
8111     In Russian Doping Scandal, Time for a Punishme...
10794         Friday Fox Follies – Megyn & Murdoch’s Money
13326    Johnny Nicholson, Whose Midtown Cafe Drew the ...
                               ...                        
742      3 Reasons Why You Should Apply For A Job In Th...
11559                           The Factless Fact-Checkers
5257     This Is Why America Will Get the President it ...
9784     Clinton Staffer Caught Sacrificing Baby to Dem...
14343                Is Western Civilization Worth Saving?
Name: title, Length: 810, dtype: object
15475    0
15645    1
8111     0
10794    1
13326    0
        ..
742      1
11559    1
5257     1
9784     1
14343    1
Name: label, Length: 810, dtype: int64


In [6]:
X_test_array = X_test.to_numpy()
Y_test_array = Y_test.to_numpy()

### Part 1: Understanding The Data

In [7]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(1,1),True)
understanding_data(P_word_real,P_word_fake,cv,3,"real")
understanding_data(P_word_real,P_word_fake,cv,3,"fake")

  Word/WordPairs
0 s
1 trump
2 hillary
  Word/WordPairs
0 new
1 york
2 times


As you can see the most appeared words in fake news is 3 words "new" "york" "times" :)
It was really amazing to have that result from all that data.

### Part 2: Implementing Naive Bayes


In this part we will se the result of using the unigram or bigram model for the naive bayes classifier

In [8]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(1,1),True)

We call our naive_bayes function by passing X_train_cv,Y_train_array and getting the trained parameters for our model. After that we can do our predictions really fast by using just those paramethers

We apply Count Vecotrizer to our X_train_cv we do not have to apply cv to others. The X_train_cv is so big that it gives Memory Error when we try to convert it to array. Therefore it stores it as a sparse matrix format

In [9]:
print("P_World_real is a vector which contains all the conditional probabilities for each word like \n[P(\"ali\"|real),P(\"ahmet\"|real),P(\"veli\"|real),P(\"hasan\"|real)]\n")
print(P_word_real)
print("\nP_World_fake is a vector which contains all the conditional probabilities for each word like \n[P(\"ali\"|fake),P(\"ahmet\"|fake),P(\"veli\"|fake),P(\"hasan\"|fake)]\n")
print(P_word_fake)
print("\nP_real is the total real cases in the all dataset:")
print(P_real)
print("\nP_real is the total real cases in the all dataset:")
print(P_fake)

P_World_real is a vector which contains all the conditional probabilities for each word like 
[P("ali"|real),P("ahmet"|real),P("veli"|real),P("hasan"|real)]

[4.08405733e-04 7.42555877e-05 7.42555877e-05 ... 7.42555877e-05
 7.42555877e-05 7.42555877e-05]

P_World_fake is a vector which contains all the conditional probabilities for each word like 
[P("ali"|fake),P("ahmet"|fake),P("veli"|fake),P("hasan"|fake)]

[3.65363537e-05 3.65363537e-05 3.65363537e-05 ... 3.65363537e-05
 3.65363537e-05 3.65363537e-05]

P_real is the total real cases in the all dataset:
26934

P_real is the total real cases in the all dataset:
27370


### Unigram

The application results are below ;

In [10]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(1,1),True)

run_test("Unigram, Stopwords: No",X_test_array,Y_test_array,P_word_real, P_word_fake, P_real, P_fake, uniq_len ,cv)

Unigram, Stopwords: No
Correct classified news:557 out of 810
Accuracy: 68.76543209876543
####################################


0

### Bigram

In the bigram approach we treat two words as a "word".

The application results are below ;

In [11]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(2,2),True)

run_test("Bigram, Stopwords: No",X_test_array,Y_test_array,P_word_real, P_word_fake, P_real, P_fake, uniq_len ,cv)

Bigram, Stopwords: No
Correct classified news:434 out of 810
Accuracy: 53.58024691358024
####################################


0

## Part 3:
### (a) Analyzing effect of the words on prediction

In this section, I did the only analysis for the binary situations because the absence words in the false news that strengthened the possibility real news, likewise other situations include cross possibility.

I tried it in both cases unigram and Bigram.

### In Unigram

In [12]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(1,1),True)

10 words whose presence most strongly predicts that the news is real.
And whose absence most strongly predicts that the news is fake.

In [13]:
understanding_data(P_word_real,P_word_fake,cv,10,"real")

  Word/WordPairs
0 s
1 trump
2 hillary
3 clinton
4 election
5 new
6 video
7 news
8 russia
9 war


10 words whose absence most strongly predicts that the news is real.
And whose presence most strongly predicts that the news is fake.

In [14]:
understanding_data(P_word_real,P_word_fake,cv,10,"fake")

  Word/WordPairs
0 new
1 york
2 times
3 s
4 breitbart
5 trump
6 donald
7 u
8 t
9 obama


### In Bigram

In [15]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(2,2),True)

10 words whose presence most strongly predicts that the news is real.
And whose absence most strongly predicts that the news is fake.

In [16]:
understanding_data(P_word_real,P_word_fake,cv,10,"real")

  Word/WordPairs
0 hillary clinton
1 donald trump
2 trump s
3 u s
4 america s
5 hillary s
6 clinton s
7 world war
8 onion america
9 finest news


10 words whose absence most strongly predicts that the news is real.
And whose presence most strongly predicts that the news is fake.

In [17]:
understanding_data(P_word_real,P_word_fake,cv,10,"fake")

  Word/WordPairs
0 new york
1 york times
2 donald trump
3 trump s
4 u s
5 briefing new
6 hillary clinton
7 evening briefing
8 white house
9 dies new


## StopWords

I already have done my analysis by not using stopwords. Lets make a analyse if we not exlude the stopwords

### Unigram

The application results are below ;

In [18]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(1,1),False)

run_test("Unigram, Stopwords: yes",X_test_array,Y_test_array,P_word_real, P_word_fake, P_real, P_fake, uniq_len ,cv)

Unigram, Stopwords: yes
Correct classified news:588 out of 810
Accuracy: 72.5925925925926
####################################


0

### Bigram

In the bigram approach we treat two words as a "word".

The application results are below ;

In [19]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(2,2),False)

run_test("Bigram, Stopwords: yes",X_test_array,Y_test_array,P_word_real, P_word_fake, P_real, P_fake, uniq_len ,cv)

Bigram, Stopwords: yes
Correct classified news:440 out of 810
Accuracy: 54.32098765432099
####################################


0

### By Including Stopwords
### (a) Analyzing effect of the words on prediction

In this section, I did the only analysis for the binary situations because the absence words in the false news that strengthened the possibility real news, likewise other situations include cross possibility.

I tried it in both cases unigram and Bigram.

### In Unigram

In [20]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(1,1),False)

10 words whose presence most strongly predicts that the news is real.
And whose absence most strongly predicts that the news is fake.

In [21]:
understanding_data(P_word_real,P_word_fake,cv,10,"real")

  Word/WordPairs
0 the
1 to
2 of
3 s
4 in
5 trump
6 a
7 and
8 on
9 for


10 words whose absence most strongly predicts that the news is real.
And whose presence most strongly predicts that the news is fake.

In [22]:
understanding_data(P_word_real,P_word_fake,cv,10,"fake")

  Word/WordPairs
0 the
1 new
2 york
3 times
4 to
5 s
6 breitbart
7 trump
8 in
9 of


### In Bigram

In [23]:
P_word_real,P_word_fake,P_real,P_fake,uniq_len,cv = naive_bayes(X_train,Y_train,(2,2),False)

10 words whose presence most strongly predicts that the news is real.
And whose absence most strongly predicts that the news is fake.

In [24]:
understanding_data(P_word_real,P_word_fake,cv,10,"real")

  Word/WordPairs
0 hillary clinton
1 comment on
2 of the
3 donald trump
4 in the
5 trump s
6 u s
7 america s
8 on the
9 and the


10 words whose absence most strongly predicts that the news is real.
And whose presence most strongly predicts that the news is fake.

In [25]:
understanding_data(P_word_real,P_word_fake,cv,10,"fake")

  Word/WordPairs
0 new york
1 the new
2 york times
3 donald trump
4 trump s
5 u s
6 briefing the
7 hillary clinton
8 in the
9 at the


### Test Results & Conclusion

Here we have all our results here

In [26]:
All_Results()

Unigram, Stopwords: No
Accuracy: 68.76543209876543
Correct classified news: 557 out of 810
####################################
Bigram, Stopwords: No
Accuracy: 53.58024691358024
Correct classified news: 434 out of 810
####################################
Unigram, Stopwords: yes
Accuracy: 72.5925925925926
Correct classified news: 588 out of 810
####################################
Bigram, Stopwords: yes
Accuracy: 54.32098765432099
Correct classified news: 440 out of 810
####################################


        Stopword-yes Stopword-no

Unigram  <strong>72.5925</strong>     68.7654

Bigram     54.3209     53.5802

| Stopword? | Yes     | No      |
|-----------|---------|---------|
| Unigram   | 72.5925 | 68.7654 |
| Bigram    | 54.3209 | 53.5802 |


As a result Unigram gave better result than Bigram and not excluding stopwords gave better result than excluding them. It can be maybe because of our training dataset is relatively small. Because of just using titles, unigram gave better results than bigram.
The best accuracy in our model is <strong>72.5925</strong>

We can see that Naive Bayes Classifier is relatively easy to implement and it can be really useful for Natural Language Processing. Even the Training part can be a very long but once we train it we can do everything by using the result probabilities

##                          $$\\Ismet\\OKATAR\\21727542$$